SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriverGameplayDebuggerContext.cpp

```diff
@@ -0,0 +1,422 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/SpatialNetDriverGameplayDebuggerContext.h"
+
+#include "EngineClasses/SpatialNetDriver.h"
+#include "EngineClasses/SpatialPackageMapClient.h"
+#include "EngineClasses/SpatialVirtualWorkerTranslator.h"
+#include "GameplayDebuggerCategoryReplicator.h"
+#include "Interop/Connection/SpatialWorkerConnection.h"
+#include "LoadBalancing/GameplayDebuggerLBStrategy.h"
+
+DEFINE_LOG_CATEGORY_STATIC(LogSpatialNetDriverGameplayDebuggerContext, Log, All);
+
+void USpatialNetDriverGameplayDebuggerContext::Enable(const SpatialGDK::FSubView& InSubView, USpatialNetDriver& NetDriver)
+{
+	if (NetDriver.GameplayDebuggerCtx != nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Error, TEXT("Enabling GDKGameplayDebugger more than once"));
+		return;
+	}
+
+	if (NetDriver.LoadBalanceStrategy == nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Error, TEXT("Enabling GDKGameplayDebugger too soon"));
+		return;
+	}
+
+	NetDriver.GameplayDebuggerCtx = NewObject<USpatialNetDriverGameplayDebuggerContext>();
+	NetDriver.GameplayDebuggerCtx->Init(InSubView, NetDriver);
+}
+
+void USpatialNetDriverGameplayDebuggerContext::Disable(USpatialNetDriver& NetDriver)
+{
+	if (NetDriver.GameplayDebuggerCtx == nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Error, TEXT("Disabling GDKGameplayDebugger before enabling it"));
+		return;
+	}
+
+	NetDriver.GameplayDebuggerCtx->Cleanup();
+	NetDriver.GameplayDebuggerCtx = nullptr;
+}
+
+void USpatialNetDriverGameplayDebuggerContext::Init(const SpatialGDK::FSubView& InSubView, USpatialNetDriver& InNetDriver)
+{
+	SubView = &InSubView;
+	NetDriver = &InNetDriver;
+
+	check(NetDriver->Connection && NetDriver->Sender);
+
+	if (NetDriver->LoadBalanceStrategy == nullptr)
+	{
+		UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Error, TEXT("Enabling GDKGameplayDebugger before LB strategy is setup"));
+		return;
+	}
+
+	LBStrategy = NewObject<UGameplayDebuggerLBStrategy>();
+	LBStrategy->Init(*this, *NetDriver->LoadBalanceStrategy);
+	NetDriver->LoadBalanceStrategy = LBStrategy;
+	NetDriver->Sender->UpdatePartitionEntityInterestAndPosition();
+
+	TSet<VirtualWorkerId> VirtualWorkerIds = LBStrategy->GetVirtualWorkerIds();
+
+	PhysicalToVirtualWorkerIdMap.Reserve(VirtualWorkerIds.Num());
+	for (const auto& VirtualWorkerId : VirtualWorkerIds)
+	{
+		const FString* PhysicalWorkerName = NetDriver->VirtualWorkerTranslator->GetPhysicalWorkerForVirtualWorker(VirtualWorkerId);
+		if (PhysicalWorkerName != nullptr)
+		{
+			PhysicalToVirtualWorkerIdMap.Add(*PhysicalWorkerName, VirtualWorkerId);
+		}
+		else
+		{
+			UE_LOG(LogSpatialNetDriverGameplayDebuggerContext, Error, TEXT("Failed to convert virtual worker to physical worker name"));
+		}
+	}
+}
+
+void USpatialNetDriverGameplayDebuggerContext::Cleanup()
```

Comment:
I do not think we ever need to support disabling this.
Maybe I added some methods in the DebugCtx, but that's mostly stale code.

Reply:
I'll leave it in for now, we can refactor/remove when the new load balancer arrives.